x/programs/runtime/runtime.go

```diff
@@ -5,178 +5,229 @@ package runtime
+func getRegisteredImportModules(importTypes []*wasmtime.ImportType) []string {
+	u := make(map[string]struct{}, len(importTypes))
+	imports := []string{}
+	for _, t := range importTypes {
+		mod := t.Module()
+		if mod == wasiPreview1ModName {
+			continue
+		}
+		if _, ok := u[mod]; ok {
+			continue
+		}
+		u[mod] = struct{}{}
+		imports = append(imports, mod)
 	}
+	return imports
+}
 
-	var api api.Function
-
-	if name == allocFnName || name == deallocFnName {
-		api = r.mod.ExportedFunction(name)
-	} else {
-		api = r.mod.ExportedFunction(utils.GetGuestFnName(name))
+func (r *WasmRuntime) Call(_ context.Context, name string, params ...uint64) ([]uint64, error) {
+	var fnName string
+	switch name {
+	case AllocFnName, DeallocFnName, MemoryFnName:
+		fnName = name
+	default:
+		// the SDK will append the guest suffix to the function name
+		fnName = name + guestSuffix
 	}
 
-	if api == nil {
+	fn := r.inst.GetFunc(r.store, fnName)
+	if fn == nil {
 		return nil, fmt.Errorf("%w: %s", ErrMissingExportedFunction, name)
 	}
 
-	result, err := api.Call(ctx, params...)
+	fnParams := fn.Type(r.store).Params()
+	if len(params) != len(fnParams) {
+		return nil, fmt.Errorf("%w for function %s: %d expected: %d", ErrInvalidParamCount, name, len(params), len(fnParams))
+	}
+
+	callParams, err := mapFunctionParams(params, fnParams)
 	if err != nil {
-		return nil, fmt.Errorf("failed to call %s: %w", name, err)
+		return nil, err
 	}
 
-	return result, nil
+	result, err := fn.Call(r.store, callParams...)
+	if err != nil {
+		return nil, fmt.Errorf("export function call failed %s: %w", name, err)
+	}
+
+	switch v := result.(type) {
+	case int32:
+		value := uint64(result.(int32))
+		return []uint64{value}, nil
+	case int64:
+		value := uint64(result.(int64))
+		return []uint64{value}, nil
+	default:
+		return nil, fmt.Errorf("invalid result type: %v", v)
+	}
 }
 
-func (r *runtime) GetGuestBuffer(offset uint32, length uint32) ([]byte, bool) {
-	// TODO: add fee
-	// r.meter.AddCost()
+func (r *WasmRuntime) Memory() Memory {
+	return NewMemory(newExportClient(r.inst, r.store))
+}
 
-	return r.mod.Memory().Read(offset, length)
+func (r *WasmRuntime) Meter() Meter {
+	return r.meter
 }
 
-// TODO: ensure deallocate on Stop.
-func (r *runtime) WriteGuestBuffer(ctx context.Context, buf []byte) (uint64, error) {
-	// TODO: add fee
-	// r.meter.AddCost()
+func (r *WasmRuntime) Stop() {
+	r.once.Do(func() {
+		r.log.Debug("shutting down runtime engine...")
+		// send immediate interrupt to engine
+		r.store.Engine.IncrementEpoch()
+		r.cancelFn()
+	})
+}
 
-	// allocate to guest and return offset
-	result, err := r.Call(ctx, allocFnName, uint64(len(buf)))
+// PreCompileWasm returns a precompiled wasm module.
+//
+// Note: these bytes can be deserialized by an `Engine` that has the same version.
+// For that reason precompiled wasm modules should not be stored on chain.
+func PreCompileWasmBytes(programBytes []byte, cfg *Config) ([]byte, error) {
```

Comment:
LOVE that this is a separate function. Provides a ton of flexibility on how this can be done (rather than relying on the engine to do all caching)